src/bst.py

```diff
@@ -0,0 +1,285 @@
+        self.length = 0
+        if root:
+            self.root = Node(root)
+            self.length = 1
+        else:
+            self.root = root
+
+    def insert(self, val, data=None):
+        """Inserts a value  into the BST.  If the value is already in
+        the BST it will be ingored."""
+        if not self.contains(val):
+            if self.root:
+                self.root._insert(val, data)
+            else:
+                self.root = Node(val, data)
+            self.length += 1
+
+    def contains(self, val):
+        """Will return True if val is in the BST, or False if it's not."""
+        try:
+            return self._contains(val, self.root)
+        except AttributeError:
+            pass
+
+    def _contains(self, val, current_node):
+        """Helper method to contains, recursively called and returns True
+        if we find the node we're looking for, False if not."""
+        if val == current_node.val:
+            return True
+        elif current_node.right and val > current_node.val:
+            return self._contains(val, current_node.right)
+        elif current_node.left and val < current_node.val:
+            return self._contains(val, current_node.left)
+        else:
+            return False
+
+    def size(self):
+        """Will return the integer size of the BST, zero if BST is empty."""
+        return self.length
+
+    def __len__(self):
+        """Returns size of tree using builtin length method."""
+        return self.length
+
+    def depth(self, starting_point=None):
+        """Will return the depth of the tree by counting "levels".  An empty
+        BST will return 0, a BST with 1 value will return 1, a BST with 2
+        values will return 2, and the value will fluctuate from then on
+        depending on the branch layout."""
+        if starting_point is None:
+            starting_point = self.root
+        # import pdb; pdb.set_trace();
+        try:
+            ld = starting_point.left.depth
+        except AttributeError:
+            ld = 0
+        try:
+            rd = starting_point.right.depth
+        except AttributeError:
+            rd = 0
+        try:
+            md = starting_point.depth
+        except AttributeError:
+            md = 0
+        return max(ld, rd, md)
+
+    def balance(self, starting_point=None):
+        """Will return an integer that's positive or negative that represents
+        the difference between depth on both sides from the root."""
+        if starting_point is None:
+            starting_point = self.root
+        return starting_point.left - starting_point.right
+
+    def in_order(self, starting_point=None):
+        """
+        This function will return a generator that will return the values
+        of the tree using in-order traversal, one value at a time.
+        """
+        if starting_point is None:
+            starting_point = self.root
+        return starting_point._in_order()
```

Comment:
Every time you write a data structure without testing every method of that data structure on an empty case, you will be subject to my bad jokes.

You have been warned.